#### The principal choice of XGBoost model:

Linear regression was chosen as a simple benchmark to highlight the benefits of more powerful models like XGBoost, which can capture non-linear patterns and temporal effects that linear regression cannot.# We're gonna choose for benchmarking and interpretability linear regression, we aren't expecting to better  model than XGBoost because it struggle on handling non linearity patterns compared to tree based model, the reason of choosing just to understand regression task 

In [1]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.metrics import  mean_squared_error, r2_score
import numpy as np

In [2]:
#Load datasets

train_hour_df = pd.read_csv("D:/VS Code Projects/Datasets/Bike Sharing/data_processed/feature_engineered_train_hour.csv")
valid_hour_df = pd.read_csv("D:/VS Code Projects/Datasets/Bike Sharing/data_processed/feature_engineered_valid_hour.csv")

In [3]:
### Ensure datetime & remove the column 'Unnamed: 0'
datasets_list =  [train_hour_df,valid_hour_df]
for dataset in datasets_list:
    dataset['date'] = pd.to_datetime(dataset['date'], errors= 'coerce')
    dataset.drop(columns=['Unnamed: 0'], inplace=True, errors='ignore')

print('ISSUE FIXED')


ISSUE FIXED


In [4]:
# ================================================
# Define target & features
# ================================================
target = 'num_rentals'
X_hour_train = train_hour_df.drop(columns= [target, 'date'])
y_hour_train = train_hour_df[target]

X_hour_valid = valid_hour_df.drop(columns= [target, 'date'])
y_hour_valid =valid_hour_df[target]

In [5]:
# ================================================
# Standardization (fit on train, transform eval)
# ================================================

scaler = StandardScaler()

X_hour_train_scaled = scaler.fit_transform(X_hour_train)
X_hour_valid_scaled = scaler.fit_transform(X_hour_valid)

In [6]:
# ================================================
# Train & Evaluate Models
# ================================================
lr =LinearRegression()


In [7]:
lr.fit(X_hour_train_scaled, y_hour_train)
y_hour_pred = lr.predict(X_hour_valid_scaled)

In [8]:
# Prediction Metrics
def evaluate (y_true, y_pred, label):
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    r2 = r2_score(y_true, y_pred)
    print(f"{label} → RMSE: {rmse:.2f}, R²: {r2:.3f}")

In [9]:
evaluate(y_hour_valid, y_hour_pred, "Hourly Linear Regression")

Hourly Linear Regression → RMSE: 123.66, R²: 0.657


The model performs decently, slightly better than the persistence baseline (R² ~0.657, RMSE ~124), but not as well as XGBoost (R² 0.957, RMSE 43.86). It likely struggles because hourly rentals exhibit non-linear patterns, which XGBoost handles very well.  

